packages/client/hmi-client/src/services/model-templating.ts

```diff
@@ -78,14 +82,66 @@ export function junctionCleanUp(modelTemplates: ModelTemplates) {
 	modelTemplates.junctions = modelTemplates.junctions.filter(({ edges }) => edges.length > 1);
 }
 
-// Designed to work with the decomposed model templates
-function appendNumberToModelVariables(model: any, number: number) {
-	const newModel = cloneDeep(model);
-	const { states, transitions } = newModel.model;
-	const { rates, initials, parameters, observables } = newModel.semantics.ode;
+/**
+ * Add/remove template cards and edges in the UI (view) and kernel are separate functions
+ * The view functions are always called at the end of the kernel functions
+ * There are cases where the view functions can be called without the kernel functions
+ */
+
+/**
+ * Add template card
+ */
+
+// Helper functions
+function determineNumberToAppend(models: any[]) {
+	// Collect values to check for numbers
+	const valuesToCheck: Set<string> = new Set();
+	models.forEach((model) => {
+		valuesToCheck.add(model.header.name);
+		valuesToCheck.add(model.metadata.templateCard.name);
+
+		const { states, transitions } = model.model;
+		const { rates, initials, parameters, observables } = model.semantics.ode;
+
+		[...states, ...transitions, ...rates, ...initials, ...parameters, ...observables].forEach(
+			(variable) => {
+				const { id, name, target, properties, input, output, expression } = variable;
+
+				// Common properties
+				if (id) valuesToCheck.add(id);
+				if (name) valuesToCheck.add(name);
+				if (target) valuesToCheck.add(target);
+				// Transition properties
+				if (properties?.name) valuesToCheck.add(properties.name);
+				if (input) input.forEach((value: string) => valuesToCheck.add(value));
+				if (output) output.forEach((value: string) => valuesToCheck.add(value));
+
+				// Check if the expression contains any variables with a number after it
+				if (expression) {
+					const matches = expression.match(/([a-zA-Z])(\d+)/g);
+					if (matches) matches.forEach((match: string) => valuesToCheck.add(match));
+				}
```

Comment:
For what it is worth, I don't think you need to check for expressions, since they are derived from parameters and initials and you included them already from above.